File: leetcode/3sum/3sum.cpp
/**************************************
* @author dujiong
* @date 2016.11.28
* @version V0.1
**************************************/

class Solution {
public:
    vector<vector<int> > threeSum(vector<int> &num) {
    	vector<vector<int> > ret;
        if(num.size()<3){
            return ret;
        }
        vector<int> temp(3,0);
        std::sort(num.begin(), num.end());
        int last=num[0];
        for(int i=0;i<num.size()-1 && num[i]<=0;i++){
			if(i && num[i]==last)
                continue;
            int left=i+1;
            int right = num.size()-1;
            while(left < right){
                int sum = num[left]+num[right];
                if(sum+num[i]==0){
                    temp[0]=num[i];
                    temp[1]=num[left];
                    temp[2]=num[right];
                    ret.push_back(temp);
                    left++;
                    right--;
                }else if(sum+num[i]<0){
                    left++;
                }else{
                    right--;
                }
            }
            last = num[i];
        }
        ret.erase(unique(ret.begin(),ret.end()), ret.end());
        return ret;
    }
};

File: leetcode-golang/3Sum/3Sum.go
import "sort"

func threeSum(nums []int) [][]int {
	var result [][]int
	sort.Ints(nums)
	for i := 0; i < len(nums)-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue //避免重复，只能避免初始点
		}
		//O(N2)的时间复杂度，两根指针去找两个数相加等于nums[i]
		target, left, right := -nums[i], i+1, len(nums)-1
		for left < right {
			if nums[left]+nums[right] == target {
				result = append(result, []int{nums[i], nums[left], nums[right]})
				left++
				right--
				//再次去重
				for left < right && nums[left] == nums[left-1] {
					left++
				}
				for left < right && nums[right] == nums[right+1] {
					right--
				}
			} else if nums[left]+nums[right] < target {
				left++
			} else {
				right--
			}
		}
	}
	return result
}